File: app/lib/core/bars.dart
import 'dart:convert';
import 'dart:math' as math;

import 'fmt.dart';
import 'indicators.dart';
import 'net.dart';

/// Daily bars for one stock (or index), from Naver, for the charts.
class Bars {
  List<String> date = [];
  List<double> open = [], high = [], low = [], close = [], volume = [];
  late Series series;

  static const _url = 'https://api.finance.naver.com/siseJson.naver?symbol=%s'
      '&requestType=1&startTime=%from&endTime=%to&timeframe=day';
  static final Map<String, Bars> _cache = {};
  static const _cacheSize = 30;

  /// Forget every stock's bars so the next expand loads today's.
  static void clear() => _cache.clear();

  static Bars? cached(String ticker) => _cache[ticker];

  static String _day(DateTime t) => '${t.year}${two(t.month)}${two(t.day)}';

  /// About 400 calendar days, enough for the 120-day average and a settled RSI.
  static Future<Bars> load(String ticker) async {
    final now = DateTime.now();
    final text = await Net.get(_url
        .replaceFirst('%s', ticker)
        .replaceFirst('%from', _day(now.subtract(const Duration(days: 420))))
        .replaceFirst('%to', _day(now)));
    // The response is almost JSON: the header row uses single quotes.
    final rows = jsonDecode(text.trim().replaceAll("'", '"')) as List;
    final b = Bars();
    for (var i = 1; i < rows.length; i++) {
      final r = rows[i] as List;
      double v(int j) => j < r.length && r[j] is num ? (r[j] as num).toDouble() : double.nan;
      final o = v(1), h = v(2), l = v(3), c = v(4), vol = v(5);
      // Days without trades carry no price information; the scan leaves them out too.
      if (!(vol > 0 && o > 0 && h > 0 && l > 0 && c > 0)) continue;
      b.date.add('${r[0]}'.trim());
      b.open.add(o);
      b.high.add(math.max(h, math.max(o, c)));
      b.low.add(math.min(l, math.min(o, c)));
      b.close.add(c);
      b.volume.add(vol);
    }
    b.series = compute(b.high, b.low, b.close);
    if (_cache.length >= _cacheSize) _cache.remove(_cache.keys.first);
    _cache[ticker] = b;
    return b;
  }

  int get size => close.length;
}

File: android/app/src/main/java/kr/personal/oscalert/Bars.java
package kr.personal.oscalert;

import android.util.LruCache;

import org.json.JSONArray;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

/** Daily bars for one stock, from Naver, for the charts. */
final class Bars {
    String[] date;
    double[] open, high, low, close, volume;
    Indicators.Series series;

    private static final String URL = "https://api.finance.naver.com/siseJson.naver?symbol=%s"
            + "&requestType=1&startTime=%s&endTime=%s&timeframe=day";
    private static final LruCache<String, Bars> CACHE = new LruCache<>(30);
    private static final long DAY = 86_400_000L;

    static Bars cached(String ticker) {
        return CACHE.get(ticker);
    }

    /** About 400 calendar days, enough for the 120-day average and a settled RSI. Blocking. */
    static Bars load(String ticker) throws Exception {
        SimpleDateFormat f = new SimpleDateFormat("yyyyMMdd", Locale.ROOT);
        long now = System.currentTimeMillis();
        String text = Net.get(String.format(Locale.ROOT, URL, ticker,
                f.format(new Date(now - 420 * DAY)), f.format(new Date(now))));
        // The response is almost JSON: the header row uses single quotes.
        JSONArray rows = new JSONArray(text.trim().replace('\'', '"'));
        List<double[]> vals = new ArrayList<>();
        List<String> dates = new ArrayList<>();
        for (int i = 1; i < rows.length(); i++) {
            JSONArray r = rows.getJSONArray(i);
            double o = r.optDouble(1), h = r.optDouble(2), l = r.optDouble(3), c = r.optDouble(4), v = r.optDouble(5);
            // Days without trades carry no price information; the scan leaves them out too.
            if (!(v > 0 && o > 0 && h > 0 && l > 0 && c > 0)) continue;
            dates.add(r.getString(0).trim());
            vals.add(new double[]{o, Math.max(h, Math.max(o, c)), Math.min(l, Math.min(o, c)), c, v});
        }
        Bars b = new Bars();
        int n = vals.size();
        b.date = dates.toArray(new String[0]);
        b.open = new double[n]; b.high = new double[n]; b.low = new double[n];
        b.close = new double[n]; b.volume = new double[n];
        for (int i = 0; i < n; i++) {
            double[] v = vals.get(i);
            b.open[i] = v[0]; b.high[i] = v[1]; b.low[i] = v[2]; b.close[i] = v[3]; b.volume[i] = v[4];
        }
        b.series = Indicators.compute(b.high, b.low, b.close);
        CACHE.put(ticker, b);
        return b;
    }

    int size() {
        return close.length;
    }
}
